packages/booter-lb3app/src/lb3app.booter.ts

```diff
@@ -0,0 +1,128 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/booter-lb3app
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {BootBindings, Booter} from '@loopback/boot';
+import {CoreBindings, inject} from '@loopback/core';
+import {
+  ExpressRequestHandler,
+  OpenApiSpec,
+  rebaseOpenApiSpec,
+  RestApplication,
+} from '@loopback/rest';
+import * as debugFactory from 'debug';
+import {Application as ExpressApplication} from 'express';
+import pEvent from 'p-event';
+import * as path from 'path';
+
+const {generateSwaggerSpec} = require('loopback-swagger');
+const swagger2openapi = require('swagger2openapi');
+
+const debug = debugFactory('loopback:boot:lb3app');
+
+const DefaultOptions: Lb3AppBooterOptions = {
+  // from "/dist/application.ts" to "/legacy"
+  path: '../legacy/server/server',
+  mode: 'fullApp',
+  restApiRoot: '/api',
+};
+
+export class Lb3AppBooter implements Booter {
+  options: Lb3AppBooterOptions;
+  appPath: string;
+
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE)
+    public app: RestApplication,
+    @inject(BootBindings.PROJECT_ROOT)
+    public projectRoot: string,
+    @inject(`${BootBindings.BOOT_OPTIONS}#lb3app`)
+    options: Partial<Lb3AppBooterOptions> = {},
+  ) {
+    this.options = Object.assign({}, DefaultOptions, options);
+
+    if (typeof app.mountExpressRouter !== 'function') {
+      throw new Error(
+        'Lb3AppBooter requires RestApplication with mountExpressRouter API',
+      );
+    }
+  }
+
+  async configure?(): Promise<void> {
+    this.appPath = path.join(this.projectRoot, this.options.path);
+  }
+
+  async load?(): Promise<void> {
+    const legacyApp = await this.loadAndBootTheApp();
+    const spec = await this.buildOpenApiSpec(legacyApp);
+    if (this.options.mode === 'fullApp') {
+      this.mountFullApp(legacyApp, spec);
+    } else {
+      this.mountRoutesOnly(legacyApp, spec);
+    }
+
+    // TODO(bajtos) Listen for the following events to update the OpenAPI spec:
+    // - modelRemoted
+    // - modelDeleted
+    // - remoteMethodAdded
+    // - remoteMethodDisabled
+    // Note: LB4 does not support live spec updates yet. See
+    // https://github.com/strongloop/loopback-next/issues/2394 for details.
+  }
+
+  private async loadAndBootTheApp() {
+    debug('Loading LB3 app from', this.appPath);
+    const legacyApp = require(this.appPath);
+
+    // NOTE(bajtos) loopback-boot@3 sets "booting" flag asynchronously,
+    // it can have one of the following values:
+    // - undefined: boot has not started yet or is in (early) progress
+    // - true: boot is in progress
+    // - false: boot was finished
+    if (legacyApp.booting !== false) {
```

Comment:
The solution I used in the spike worked well for loopback-boot@3 but unfortunately did not work for loopback-boot@2 :(

I have recently fixed this problem in `loopback-boot@3.2.1`, see https://github.com/strongloop/loopback-boot/pull/286. Now we should have the same behavior in both version of loopback-boot, and the check here should rewritten as follows:

```suggestion
    if (legacyApp.booting) {
```